lib/de.ml

```diff
@@ -1445,6 +1448,284 @@ module Inf = struct
+        | Invalid_dictionary -> "Invalid dictionary"
+        | Invalid_complement_of_length -> "Invalid complement of length"
+        | Invalid_distance -> "Invalid distance"
+        | Invalid_distance_code -> "Invalid distance code" in
+      Format.fprintf ppf "%s" s
+
+    exception Malformed of error
+
+    let err_unexpected_end_of_input () =
+      raise (Malformed Unexpected_end_of_input)
+
+    let err_unexpected_end_of_output () =
+      raise (Malformed Unexpected_end_of_output)
+
+    let err_invalid_kind_of_block () = raise (Malformed Invalid_kind_of_block)
+    let err_invalid_dictionary () = raise (Malformed Invalid_dictionary)
+
+    let err_invalid_complement_of_length () =
+      raise (Malformed Invalid_complement_of_length)
+
+    let err_invalid_distance () = raise (Malformed Invalid_distance)
+    let err_invalid_distance_code () = raise (Malformed Invalid_distance_code)
+
+    (* remaining bytes to read [d.i]. *)
+    let i_rem d = d.i_len - d.i_pos [@@inline]
+
+    let flat d =
+      d.i_pos <- d.i_pos - (d.bits / 8)
+      ; d.hold <- 0
+      ; d.bits <- 0
+      ; if i_rem d < 4 then err_unexpected_end_of_input ()
+      ; let len = unsafe_get_uint16 d.i d.i_pos in
+        let nlen = unsafe_get_uint16 d.i (d.i_pos + 2) in
+        d.i_pos <- d.i_pos + 4
+        ; if nlen != 0xffff - len then err_invalid_complement_of_length ()
+          else (
+            if len > i_rem d then err_unexpected_end_of_input ()
+            ; if len > d.o_len - d.o_pos then err_unexpected_end_of_output ()
+            ; WInf.blit d.w d.i d.i_pos d.o d.o_pos len
+            ; d.o_pos <- d.o_pos + len
+            ; d.i_pos <- d.i_pos + len)
+
+    let _fill_bits d =
+      if i_rem d < 1 then err_unexpected_end_of_input ()
+      else if i_rem d <> 1 then (
+        d.hold <- d.hold lor (unsafe_get_uint16 d.i d.i_pos lsl d.bits)
+        ; d.i_pos <- d.i_pos + 2
+        ; d.bits <- d.bits + 16)
+      else (
+        d.hold <- d.hold lor (unsafe_get_uint8 d.i d.i_pos lsl d.bits)
+        ; d.i_pos <- d.i_pos + 1
+        ; d.bits <- d.bits + 8)
+      [@@inline]
+
+    (* clecat: Removed the rec flag on fill_bits, as it was never called with
+       d greater than 16. However, we should make sure that it is never done in
+       the future. *)
+    let fill_bits d n = if d.bits < n then _fill_bits d
+
+    let pop_bits d n =
+      let v = d.hold land ((1 lsl n) - 1) in
+      d.hold <- d.hold lsr n
+      ; d.bits <- d.bits - n
+      ; v
+      [@@inline]
+
+    exception End
+    exception Invalid_distance
+    exception Invalid_distance_code
+
+    let inflate lit dist d =
+      let fill_bits d lit n =
+        if
+          d.bits < n
+          && not
+               (lit.Lookup.t.(d.hold land lit.Lookup.m) land Lookup.mask == 256
+               && lit.Lookup.t.(d.hold land lit.Lookup.m) lsr 15 <= d.bits
+               && i_rem d < 1)
+        then _fill_bits d in
```

Comment:
May be inline `fill_bits` too, not sure that it's the right solution however.

Reply:
After checking with landmarks, it does not seem to be any better